x/programs/wasmlanche/src/host.rs

```diff
@@ -0,0 +1,284 @@
+    }
+
+    impl Accessor {
+        pub fn new() -> Self {
+            Accessor {
+                state: MockState::new(),
+            }
+        }
+
+        pub fn state(&self) -> &MockState {
+            &self.state
+        }
+
+        pub fn new_deploy_address(&self) -> Address {
+            let address: [u8; 33] = [self.state().deploy(); 33];
+            Address::new(address)
+        }
+
+        pub fn deploy(&self, key: &[u8]) -> HostPtr {
+            let key = [DEPLOY_PREFIX]
+                .iter()
+                .chain(key.iter())
+                .copied()
+                .collect::<Vec<u8>>();
+            let val = self.state.get(&key);
+
+            assert!(
+                !val.is_null(),
+                "Deploy function not mocked. Please mock the function call."
+            );
+
+            val
+        }
+
+        pub fn call_program(&self, args: &[u8]) -> HostPtr {
+            let key = [CALL_FUNCTION_PREFIX]
+                .iter()
+                .chain(args.iter())
+                .copied()
+                .collect::<Vec<u8>>();
+            let val = self.state.get(&key);
+
+            assert!(
+                !val.is_null(),
+                "Call function not mocked. Please mock the function call."
+            );
+
+            val
+        }
+
+        pub fn get_balance(&self, args: &[u8]) -> HostPtr {
+            // balance prefix + key
+            let key = [BALANCE_PREFIX]
+                .iter()
+                .chain(args.iter())
+                .copied()
+                .collect::<Vec<u8>>();
+
+            let host_ptr = self.state.get(&key);
+            assert!(
+                !host_ptr.is_null(),
+                "get_balance not mocked. Please mock the function call."
+            );
+
+            host_ptr
+        }
+
+        pub fn get_remaining_fuel(&self) -> HostPtr {
+            self.state().get_fuel()
+        }
+
+        pub fn send_value(&self, args: &[u8]) -> HostPtr {
+            // send prefix + key
+            let key = [SEND_PREFIX]
+                .iter()
+                .chain(args.iter())
+                .copied()
+                .collect::<Vec<u8>>();
+
+            let host_ptr = self.state.get(&key);
+            assert!(
+                !host_ptr.is_null(),
+                "send_value not mocked. Please mock the function call."
+            );
+
+            host_ptr
+        }
+    }
+
+    impl Default for MockState {
+        fn default() -> Self {
+            Self::new()
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct MockState {
+        state: RefCell<hashbrown::HashMap<Vec<u8>, Vec<u8>>>,
+        deploys: RefCell<u8>,
+        fuel: Vec<u8>,
+    }
+
+    impl MockState {
+        pub fn new() -> Self {
+            Self {
+                state: RefCell::new(hashbrown::HashMap::new()),
+                deploys: RefCell::new(0),
+                fuel: borsh::to_vec(&u64::MAX).expect("failed to serialize args"),
```

Comment:
store as a vec so `HostPtr` always references valid memory

Reply:
?

Reply:
not relevant anymore, but was doing weird things for no reason